crypto/evp/p_lib.c

```diff
@@ -347,7 +373,10 @@ int EVP_PKEY_eq(const EVP_PKEY *a, const EVP_PKEY *b)
     if (a == NULL || b == NULL)
         return 0;
 
-    if (a->keymgmt != NULL || b->keymgmt != NULL) {
+#ifndef FIPS_MODULE
+    if (a->keymgmt != NULL || b->keymgmt != NULL)
+#endif  /* !FIPS_MODULE */
```

Comment:
Not sure I understand this.  Is the keymgmt pointer for a and b guaranteed to be non-null in the FIPS provider?  Would it just be better to make this check universally?  If that guarantee is in place, then this seems fine, but it might make the code less convoluted to just eliminate the ifdef

Reply:
Yes, they are guaranteed to be non-null. If we would keep this if() we would also need to keep `return -2;` below.

Reply:
@slontis bought up a similar point [above](https://github.com/openssl/openssl/pull/26714#discussion_r1953491969).  The FIPS provider omits **all** legacy support, so no need for the checks.  I adopted the pattern of conditioning out all such checks.  As noted by @t8m, leaving these here (rightly) confuses compilers and forces the addition of extra code to placate them.

Once legacy support is removed (I'm praying for this in 4.0), a **lot** of mess will disappear.  Doing it this way makes such removal fractionally easier.

Still, if someone insists, I'm okay to leave these checks in and include the extra returns etc.  I don't see much of a point and the clutter will hopefully be temporary.